README

Author: Ulric Eriksson (ulric@siag.nu)

This is libsdb, the simple database library, which provides a way
to support multiple database management systems in an application
with negligeable overhead, in terms of code as well as system
resources.

Supported databases:

 - Gdbm
 - Ingres
 - Lago
 - Microsoft SQL (requires freetds >= 0.64)
 - Mimer
 - Msql
 - Mysql
 - ODBC
 - Oracle
 - Postgresql
 - Sdbd (libsdb proxy)
 - Shsql/quisp
 - Sqlite versions 2 and 3
 - Sybase (untested, see Microsoft SQL)
 - Text

The driver for ODBC can be used for any database
supported by UnixODBC, including Sybase. It can also
be on Windows for access to any database which has a
Windows ODBC driver.

Gdbm does not use SQL. See gdbm.c for usage.


Installation
------------

See INSTALL for installation instructions.


Database interface
------------------

The database interface is a single function, sdb_query. Synopsis:

        int sdb_query(char *url, char *query,
                int (*callback)(int, char **, void *), void *closure)

url is a string which identifies the driver and any necessary options
such as the name of the database and user. The set of options is
driver specific. Examples:

	ingres:db=XXX
        mysql:host=XXX:db=XXX:uid=XXX:pwd=XXX
        postgres:host=XXX:db=XXX:port=XXX
        oracle:uid=XXX:pwd=XXX OR oracle:uid=username/password@instance
        sqlite:db=/full/path/to/db
        sqlite3:db=/full/path/to/db
        lago:host=XXX:port=XXX:db=XXX:uid=XXX:pwd=XXX
        mimer:db=XXX:uid=XXX:pwd=XXX
        odbc:dsn=postgresql
	sdbd:host:port:url=XXX
	shsql: (path to db in SHSQL_DB environment variable)
	tds:host=XXX:port=XXX:uid=XXX:pwd=XXX
	text:/path/to/db/ (note trailing slash)

Some drivers also get parameters from environment variables, such
as ORACLE_SID.

query is an SQL string sent to the database, such as

        select count(*) from crontab where host='myhost'

sdb_query calls the callback once for each row returned. No rows does
not necessarily indicate an error condition; for example

        delete from crontab where host='myhost'

does not return anything. sdb_query returns the number of rows or -1
for error.

The callback takes three arguments, an integer indicating the number
of columns in the reply, an array of pointers to the fields and a
pointer to some arbitrary data that the callback might need.
Values are always returned as strings. Synopsis:

        int callback(int n, char **p, void *closure)

The file sdb_client.c contains a very simple line-mode client. Example:

sh-2.03$ ./sdb_client oracle:uid=system/manager@ulric
sdb> select * from counters
serialno 1 
serial 17 
Return code: 2
sdb> update counters set value=2 where name='serialno'
Return code: 0
sdb> select * from counters
serialno 2 
serial 17 
Return code: 2
sdb>


Proxy daemon
------------

Sometimes linking directly with the library isn't practical,
perhaps because an application is being distributed in binary form
and there's no way to know what database will be used, or perhaps
the application will be run on a different machine from where the
database client libraries are installed.

Sdbd to the rescue! This is a proxy daemon which handles the database
communication on behalf of the application. As a result, there is
no need for the application to be linked to database libraries or even
libsdb itself; all that is handled by sdbd.

An example client is provided in sdbd_client.c. It is very similar
to sdb_client, except that it doesn't use any libsdb calls. Example:

ulric@host1$ sdbd 2222

ulric@host2$ sdbd_client host1 2222 postgres:host=host1:db=testdb
sdbd> select * from foo
Matilda Eriksson 
Ulric Eriksson 
Joel Eriksson 
sdbd> 

In this case, the database, the client libraries and libsdb are all
installed on host1. Nothing except sdbd_client is needed on host2.

Description of the protocol, as seen on the wire:

1. Client sends url
2. Server sends single space as acknowledgement
3. Client sends query string
4. Server sends number of columns in result, space-terminated
5. Server sends column size in bytes, space-terminated
6. Server sends column contents
7. 5-6 are repeated for all additional columns
8. 4-7 are repeated for all additional result rows
9. Server sends the number 0.

Everything is sent in ascii form, so it can be debugged by humans.
Example:

sh-2.05$ telnet localhost 2222
Trying 127.0.0.1...
Connected to localhost.
Escape character is '^]'.
postgres:host=localhost:db=testdb
 select * from foo
2 7 Matilda8 Eriksson2 5 Ulric8 Eriksson2 4 Joel8 Eriksson0
Connection closed by foreign host.



ChangeLog

150222	Updated config.in.

101005	More tweaking of the configuration script for compatibility
	with the latest instant oracle client.
	Released 0.10.1.

080430	Oracle driver again: increased the size of the hda to 512 bytes
	for compatibility with 64-bit platforms.
	Released 0.10.0.

080427	Made the Oracle driver build compatible with Oracle Instant Client.

080425	Added db option to oracle url. These are now equivalent:
	oracle:uid=username/password@database
	oracle:uid=username:pwd=password:db=database

080311	Added --disable-shared and --disable-static options to
	the configure script.
	Released 0.9.0.

080307	In sdb.c, function sdb_close: don't call the driver's close
	function if the db is null (i.e. there is nothing to close).

080306	Started integrating Freetds for Microsoft SQL and Sybase.
	Queries are submitted but returned values are not handed
	off to the callback.

080305	Oracle driver updated for 11g.
	Make sure that the Oracle driver treats null values as NULL.
	Released 0.8.0.

080304	Initial support for Ingres.

080229	Added support for shsql/quisp.
	Changed clilen type to socklen_t in sdbd.c.
	Released 0.7.0.

051029	Added -c option to sdb_client to read an sql query from the
	command line rather than interactively.
	Released 0.6.2.

050219	In sdb_query: if the url refers to an already opened connection,
	check that the connection is in fact opened. This is important
	for mistyped urls.
	Released 0.6.1.

050110	Added --platform flag to the configure script.

050110	Added support for Sqlite 3.
	Added another minimal example to the manpage.
	Released 0.6.0.

050109	Dumped automake. Its self-incompatibility drives me nuts!

040506	More bugfixes from Guillaume Cottenceau to the Postgres
	and Mysql drivers: don't segfault if sdb_close is called
	with a NULL argument.
	Released 0.5.5.

040423	Bugfix in the postgres driver: the closer should be called
	sdb_postgres_close. Released 0.5.4.

040421	Patch for persistent Postgres connections from
	Guillaume Cottenceau <gc at mnc.ch>. Released 0.5.3.

040227	Check for Solaris specific libraries.
	Added patch for persistent Oracle connections
	from Denis Ovsienko <pilot@altlinux.ru>.

020828	Do not link sdb_client against .libs/libsdb.so. It breaks the
	build on certain platforms.
	sdb_client: only print prompt if stdin is a terminal.
	sdb_client: Option -s changes the separator in output.
	Released 0.5.1.

020630	Added a driver by Lahcen EL HADDATI <elhaddati@hotmail.com>
	which accepts queries on plain text files in a SQL-like language.
	Released 0.5.0.

020621	Fixed bug in sdbd: null values in the result set caused the
	process to crash. Now it returns an empty string, which is wrong
	but better than crashing.
	Check memory allocations better in sdbd_driver.

020620	Mimer driver that actually works.
	Released 0.4.1.

020530	MiniSQL driver. Use --with-msql=/path/to/msql if it is not
	installed in /usr/local/Hughes.
	Released 0.4.0.

020526	Implemented open and close for mysql driver.
	Wrote manpages for sdb_open, sdb_close, sdb_query, sdb_init and sdbd.
	Released 0.3.0.

020522	Added sdb_open and sdb_close for connections that are to be
	used for more than one query.
	Implemented open and close for sqlite driver.
	Modified the sdbd protocol to allow connections to be used for
	multiple queries. Updated sdbd.c and sdbd_driver.c to use the new
	protocol if it is implemented on both server and client.

020520	Added a check in sdb_query to call sdb_init if it hasn't already
	been done. This makes sdb_init calls from applications optional.

020515	Improved error checking in gdbm_driver.c.

020514	Added squid_sdb_auth.c, a program to authenticate Squid proxy
	users to any database supported by libsdb.
	Released 0.2.4.

020503	Renamed all the drivers from XXX.c to XXX_driver.c.
	Wrote a new driver (sdbd_driver.c) for proxied queries.
	The new driver is designed so it can be compiled statically
	into an application without the rest of libsdb.

020429	Added optional argument to --with-odbc option, plus the necessary
	tweaking to get the odbc driver running on Windows.
	Changed #include <mysql/mysql.h> to just #include <mysql.h>.
	Released 0.2.3.

020425	New flags -d and -f to sdbd control debug message level and
	background/foreground operation.
	Made sdbd less eager to error out at every problem.

020209	Modification to postgres driver to let it work with
	user/password login (srosso@libero.it).
	Released 0.2.2.

020105	Fixed compiler warning in sdbd_client.c (2nd arg to connect).

011220	Updated automake, autoconf and libtool.

011204	Updated configuration for Mysql.

010910	SO_REUSEADDR in sdbd.c.

010611	Fixed silly bug in sdbd.c and sdbd_client.c: readdata and writedata
	would fail when data was read and written in small segments.
	Released 0.2.1.

010610	Sample client for sdbd in sdbd_client. Released 0.2.0.

010601	Proxy daemon called sdbd.

010520	Updated documentation and sdb_client.
	Released 0.1.0.

010517	Added trivial gdbm driver (adapted from Siag).

010515	Added autoconf and libtool.

010513	Released 0.0.1.

010512	Cleaned up namespace by prefixing all globals with sdb_.

010511	Started project by copying database code from Ucronadm.



tds_driver.c
/*
   Copyright (C) 2008  Ulric Eriksson <ulric@siag.nu>

   Based on tsql.c from freetds:
   FreeTDS - Library of routines accessing Sybase and Microsoft databases
   Copyright (C) 1998, 1999, 2000, 2001, 2002, 2003, 2004, 2005 Brian Bruns

   This library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Library General Public
   License as published by the Free Software Foundation; either
   version 2 of the Licence, or (at your option) any later version.

   This library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Library General Public License for more details.

   You should have received a copy of the GNU Library General Public
   License along with this library; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place - Suite 330, Boston,
   MA 02111-1307, USA.
 */

#ifdef HAVE_LIBTDS

# include <sys/time.h>
# include <time.h>

#include <stdio.h>
#include <assert.h>
#include <ctype.h>

#include <errno.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <locale.h>

#ifdef HAVE_LANGINFO_H
#include <langinfo.h>
#endif /* HAVE_LANGINFO_H */

#include <tds.h>
#include <tdsconvert.h>
#include "common.h"
#include "sdb.h"

enum
{
	OPT_VERSION =  0x01,
	OPT_TIMER =    0x02,
	OPT_NOFOOTER = 0x04,
	OPT_NOHEADER = 0x08,
	OPT_QUIET =    0x10
};

struct conn_info {
	TDSSOCKET *tds;
	TDSLOGIN *login;
	TDSCONTEXT *context;
	TDSCONNECTION *connection;
};

#define QUIET (sdb_debuglevel == 0)

static int sdb_handle_message(const TDSCONTEXT * context, TDSSOCKET * tds, TDSMESSAGE * msg)
{
	if (msg->msgno == 0) {
		sdb_debug("%s", msg->message);
		return 0;
	}

	if (msg->msgno != 5701 && msg->msgno != 5703
	    && msg->msgno != 20018) {
		sdb_debug("Msg %d, Level %d, State %d, Server %s, Line %d\n%s",
			msg->msgno, msg->severity, msg->state, msg->server, msg->line_number, msg->message);
	}

	return 0;
}

static int do_query(TDSSOCKET* tds, char *buf, int opt_flags,
		int (*callback)(int, char **, void *), void *closure)
{
	int rows = -1;
	int rc, i;
	TDSCOLUMN *col;
	int ctype;
	CONV_RESULT dres;
	unsigned char *src;
	TDS_INT srclen;
	TDS_INT resulttype;
	struct timeval start, stop;
	char message[128];
	char **p;

	rc = tds_submit_query(tds, buf);
	if (rc != TDS_SUCCEED) {
		sdb_debug("tds_submit_query() failed");
		return -1;
	}

	while ((rc = tds_process_tokens(tds, &resulttype, NULL, TDS_TOKEN_RESULTS)) == TDS_SUCCEED) {
		if (opt_flags & OPT_TIMER) {
			gettimeofday(&start, NULL);
		}
		switch (resulttype) {
		case TDS_ROWFMT_RESULT:
			if (sdb_debuglevel && tds->current_results) {
				for (i = 0; i < tds->current_results->num_cols; i++) {
					sdb_debug("Column %d = '%s'", i, tds->current_results->columns[i]->column_name);
				}
			}
			break;
		case TDS_COMPUTE_RESULT:
		case TDS_ROW_RESULT:
			rows = 0;
			p = sdb_malloc(tds->current_results->num_cols * sizeof *p);

			while ((rc = tds_process_tokens(tds, &resulttype, NULL, TDS_STOPAT_ROWFMT|TDS_RETURN_DONE|TDS_RETURN_ROW|TDS_RETURN_COMPUTE)) == TDS_SUCCEED) {
				if (resulttype != TDS_ROW_RESULT && resulttype != TDS_COMPUTE_RESULT)
					break;

				rows++;

				if (!tds->current_results)
					continue;

				for (i = 0; i < tds->current_results->num_cols; i++) {
					col = tds->current_results->columns[i];
					if (col->column_cur_size < 0) {
						if (sdb_debuglevel)
							sdb_debug("[%d,%d] %s", rows, i, "NULL");
						p[i] = NULL;
						continue;
					}
					ctype = tds_get_conversion_type(col->column_type, col->column_size);

					src = &(tds->current_results->current_row[col->column_offset]);
					if (is_blob_type(col->column_type))
						src = (unsigned char *) ((TDSBLOB *) src)->textvalue;
					srclen = col->column_cur_size;


					if (tds_convert(tds->tds_ctx, ctype, (TDS_CHAR *) src, srclen, SYBVARCHAR, &dres) < 0)
						continue;
					if (sdb_debuglevel)
						sdb_debug("[%d,%d] %s", rows, i, dres.c);
					p[i] = sdb_strdup(dres.c);
					free(dres.c);
				}
				if (sdb_debuglevel)
					sdb_debug("Call the callback here");
				(*callback)(tds->current_results->num_cols, p, closure);
				for (i = 0; i < tds->current_results->num_cols; i++) {
					sdb_free(p[i]);
				}
			}
			sdb_free(p);
			break;
		case TDS_STATUS_RESULT:
			sdb_debug("(return status = %d)", tds->ret_status);
			break;
		default:
			break;
		}

		if (opt_flags & OPT_VERSION) {
			char version[64];
			int line = 0;

			line = tds_version(tds, version);
			if (line) {
				sprintf(message, "using TDS version %s", version);
				tds_client_msg(tds->tds_ctx, tds, line, line, line, line, message);
			}
		}
		if (opt_flags & OPT_TIMER) {
			gettimeofday(&stop, NULL);
			sprintf(message, "Total time for processing %d rows: %ld msecs\n",
				rows, (long) ((stop.tv_sec - start.tv_sec) * 1000) + ((stop.tv_usec - start.tv_usec) / 1000));
			tds_client_msg(tds->tds_ctx, tds, 1, 1, 1, 1, message);
		}
	}
	return rows;
}

static void populate_login(TDSLOGIN *login, char *url)
{
	char *hostname = sdb_url_value(url, "host");
	char *servername = sdb_url_value(url, "server");
	char *username = sdb_url_value(url, "uid");
	char *password = sdb_url_value(url, "pwd");
	char *confile = sdb_url_value(url, "cfg");
	char *portname = sdb_url_value(url, "port");
	int port = 1433;
	const char *locale = NULL;
	char *charset = NULL;

	if (portname) port = atoi(portname);

	setlocale(LC_ALL, "");
	locale = setlocale(LC_ALL, NULL);

#if HAVE_LOCALE_CHARSET
	charset = locale_charset();
#endif
#if HAVE_NL_LANGINFO && defined(CODESET)
	if (!charset)
		charset = nl_langinfo(CODESET);
#endif

	if (locale)
		if (!QUIET) sdb_debug("locale is \"%s\"", locale);
	if (charset) {
		if (!QUIET) sdb_debug("locale charset is \"%s\"", charset);
	} else {
		charset = "ISO-8859-1";
		if (!QUIET) sdb_debug("using default charset \"%s\"", charset);
	}

	/* validate parameters */
	if (!servername && !hostname) {
		sdb_error("Missing server or host");
	}
	if (hostname && !port) {
		sdb_error("Missing port");
	}
	if (!username) {
		sdb_error("Missing user");
	}
	if (!password) {
		sdb_error("Missing password");
	}

	/* all validated, let's do it */

	/* if it's a servername */

	if (servername) {
		tds_set_user(login, username);
		tds_set_app(login, "TSQL");
		tds_set_library(login, "TDS-Library");
		tds_set_server(login, servername);
		tds_set_client_charset(login, charset);
		tds_set_language(login, "us_english");
		tds_set_passwd(login, password);
		if (confile) {
			tds_set_interfaces_file_loc(confile);
		}
		/* else we specified hostname/port */
	} else {
		tds_set_user(login, username);
		tds_set_app(login, "TSQL");
		tds_set_library(login, "TDS-Library");
		tds_set_server(login, hostname);
		tds_set_port(login, port);
		tds_set_client_charset(login, charset);
		tds_set_language(login, "us_english");
		tds_set_passwd(login, password);
	}

	sdb_free(hostname);
	sdb_free(servername);
	sdb_free(username);
	sdb_free(password);
	sdb_free(confile);
	sdb_free(portname);
}

static void *sdb_tds_open(char *url)
{
	struct conn_info *ci = sdb_malloc(sizeof *ci);

	if (sdb_debuglevel) sdb_debug("sdb_tds_open(%s)", url);
	/* grab a login structure */
	ci->login = tds_alloc_login();

	ci->context = tds_alloc_context(NULL);
	if (ci->context->locale && !ci->context->locale->date_fmt) {
		/* set default in case there's no locale file */
		ci->context->locale->date_fmt = strdup("%b %e %Y %I:%M%p");
	}

	ci->context->msg_handler = sdb_handle_message;
	ci->context->err_handler = sdb_handle_message;

	/* process all the command line args into the login structure */
	populate_login(ci->login, url);

	/* Try to open a connection */
	ci->tds = tds_alloc_socket(ci->context, 512);
	tds_set_parent(ci->tds, NULL);
	ci->connection = tds_read_config_info(NULL, ci->login, ci->context->locale);
	if (!ci->connection || tds_connect(ci->tds, ci->connection) == TDS_FAIL) {
		tds_free_socket(ci->tds);
		tds_free_connection(ci->connection);
		sdb_debug("There was a problem connecting to the server");
		sdb_free(ci);
		return NULL;
	}
	tds_free_connection(ci->connection);
	return ci;
}

static int sdb_tds_close(void *db)
{
	struct conn_info *ci = db;

	if (sdb_debuglevel) sdb_debug("sdb_tds_close(%p)", db);

	/* close up shop */
	tds_free_socket(ci->tds);
	tds_free_login(ci->login);
	tds_free_context(ci->context);

	sdb_free(ci);
	return 0;
}

/* Parameters:
   pdb = pointer to previously opened database connection, or NULL
   url = url string
   query = query string
   callback = function to send the data to
   closure = pointer to data passed from application
*/
static int tds_driver(void *pdb, char *url, char *query,
	int (*callback)(int, char **, void *), void *closure)
{
	struct conn_info *ci;
	int opt_flags = 0;
	int rows;

	if (sdb_debuglevel) sdb_debug("tds_driver(%p, %s, %s, %p, %p)",
				pdb, url, query, callback, closure);
	if (pdb == NULL) ci = sdb_tds_open(url);
	else ci = pdb;
	if (ci == NULL) {
		sdb_debug("Can't open database");
		return -1;
	}

	rows = do_query(ci->tds, query, opt_flags, callback, closure);

	if (pdb == NULL) sdb_tds_close(ci);

	return rows;
}

void sdb_init_tds(void)
{
	sdb_register_driver("tds", tds_driver,
		sdb_tds_open, sdb_tds_close);
}

#else

void sdb_init_tds(void)
{
	;
}

#endif	/* HAVE_LIBTDS */
